### PR TITLE
hotfix: register "vertex" fields only in unstructured meshes

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

fix https://github.com/yt-project/yt/issues/2430

This is a hotfix, I'm convinced there are better ways to solve this but they would require much more work (refactoring).

## PR Checklist

- [x] pass `black --check yt/`
- [x] pass `isort . --check --diff`
- [x] pass `flake8 yt/`
- [x] pass `flynt yt/ --fail-on-change --dry-run -e yt/extern`
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/fields/field_info_container.py [hunk 1]
```diff
@@ -469,8 +471,9 @@ def check_derived_fields(self, fields_to_check=None):
         for field in fields_to_check:
             fi = self[field]
             try:
+                # fd: field detector
                 fd = fi.get_dependencies(ds=self.ds)
-            except Exception as e:
+            except (YTFieldNotFound, NeedsConfiguration, UnitConversionError) as e:
```
Comment (neutrinoceros) [inline yt/fields/field_info_container.py:476]: this is not strictly needed here, but as a side effect of exploring this function I figured I might as well catch relevant exceptions only. If I've missed any, tests should tell !

**Comment by neutrinoceros** [comment]
added a test that fails on master and passes here

**Comment by neutrinoceros** [comment]
switching to draft since #2850 will make this a tiny bit simpler.

**Comment by neutrinoceros** [comment]
So I think I'm going to split this between the hotfix and the experimental part where I build a long exception list to avoid `except Exception`

**Comment by neutrinoceros** [comment]
moved the exception expansion there #2851

**Comment by matthewturk** [comment]
Hm, I guess I replied over email but it didn't get through.  Anyway, I think there's a possibility we could have vertex-centered fields in some grid calculations.  Does this preclude that?

**Comment by neutrinoceros** [comment]
I don't think so. My understanding is that vertex fields only make sense for data where `ds.index.meshes` is defined. If I'm wrong then this needs to be updated of course.
More broadly, I think since is one more case where having specialised abstract Dataset subclasses could help.

**Comment by matthewturk** [comment]
@neutrinoceros As long as this does not *preclude* vertex centered fields (which might be handled with some weird logic at present) for grid fields, this looks good to me and I think it can go in.

**Review by matthewturk — APPROVED** [review]

**Comment by neutrinoceros** [comment]
I don't really know how to test this, or how to demonstrate this is fine

**Comment by matthewturk** [comment]
OK, let's hold off, it may not be the case that there are any grid
frontends to be worried about at the moment with this issue.

File: yt/fields/field_info_container.py [hunk 2]
```diff
@@ -493,6 +494,34 @@ def check_derived_fields(self, fields_to_check=None):
...
+        # now populate the derived field list with results
+        # this violates isolation principles and should be refactored
         dfl = set(self.ds.derived_field_list).union(deps.keys())
-        self.ds.derived_field_list = list(sorted(dfl, key=tupleize))
+        dfl = list(sorted(dfl, key=tupleize))
+
+        if not hasattr(self.ds.index, "meshes"):
+            # the meshes attribute characterizes a unstructured-mesh data structure
+
+            # ideally this fitlering should not be required
```
Comment (cphyc) [inline yt/fields/field_info_container.py:506]:
```suggestion
            # the meshes attribute characterizes an unstructured-mesh data structure

            # ideally this filtering should not be required
```

**Review by cphyc — APPROVED** [review]

File: yt/fields/field_info_container.py (hunk 2 above)
Comment (neutrinoceros) [reply yt/fields/field_info_container.py:506]: thanks for your careful review !

**Pull request merged**